### ISSUE TITLE
Brigade not deleting PVC after successful build

### ISSUE THREAD

**Issue by bnadeem** [brigadecore/brigade#1007]
## Problem 
While trying to use brigade storage in GKE and using FileStore of GCP for ReadWriteMany, things work fine, however, PVC is not deleted for the next iteration of the build.  

## Important observation:
The problem is that the pods of the jobs using the PVC are in `succeeded` state and is preventing the pvc to be removed. Once I manually delete the succeeded pods, the PVC is removed and the PV is ready for next build.  

## brigade.js:

```javascript
const { events, Job, Group } = require("brigadier")

events.on("exec", (e, p) => {
  var dest = "/mnt/brigade/share/hello.txt"
  var one = new Job("one", "alpine:3.4", ["echo hello > " + dest])
  var two = new Job("two", "alpine:3.4", ["echo world >> " + dest])
  var three = new Job("three", "alpine:3.4", ["cat " + dest])

  one.storage.enabled = true
  two.storage.enabled = true
  three.storage.enabled = true

  Group.runEach([one, two, three])
})
```

## output from running the the script:

```
[brigade] brigade-worker version: 1.2.1
[brigade:k8s] Creating PVC named brigade-worker-01drs34caxbnmx2b8keafw9aqp
[brigade:k8s] Creating secret one-01drs34caxbnmx2b8keafw9aqp
[brigade:k8s] Creating pod one-01drs34caxbnmx2b8keafw9aqp
[brigade:k8s] Timeout set at 900000 milliseconds
[brigade:k8s] Pod not yet scheduled
[brigade:k8s] default/one-01drs34caxbnmx2b8keafw9aqp phase Pending
[brigade:k8s] default/one-01drs34caxbnmx2b8keafw9aqp phase Succeeded
[brigade:k8s] Creating secret two-01drs34caxbnmx2b8keafw9aqp
[brigade:k8s] Creating pod two-01drs34caxbnmx2b8keafw9aqp
[brigade:k8s] Timeout set at 900000 milliseconds
[brigade:k8s] Pod not yet scheduled
[brigade:k8s] default/two-01drs34caxbnmx2b8keafw9aqp phase Pending
[brigade:k8s] default/two-01drs34caxbnmx2b8keafw9aqp phase Succeeded
[brigade:k8s] Creating secret three-01drs34caxbnmx2b8keafw9aqp
[brigade:k8s] Creating pod three-01drs34caxbnmx2b8keafw9aqp
[brigade:k8s] Timeout set at 900000 milliseconds
[brigade:k8s] Pod not yet scheduled
[brigade:k8s] default/three-01drs34caxbnmx2b8keafw9aqp phase Pending
[brigade:k8s] default/three-01drs34caxbnmx2b8keafw9aqp phase Succeeded
[brigade:app] after: default event handler fired
[brigade:app] beforeExit(2): destroying storage
[brigade:k8s] Destroying PVC named brigade-worker-01drs34caxbnmx2b8keafw9aqp
```

## PV YAML:

```yaml
apiVersion: v1
kind: PersistentVolume
metadata:
  annotations:
    kubectl.kubernetes.io/last-applied-configuration: |
      {"apiVersion":"v1","kind":"PersistentVolume","metadata":{"annotations":{},"name":"fileserver"},"spec":{"accessModes":["ReadWriteMany"],"capacity":{"storage":"1T"},"nfs":{"path":"/brigadestorage","server":"10.9.92.18"},"storageClassName":"standard"}}
    pv.kubernetes.io/bound-by-controller: "yes"
  creationTimestamp: "2019-11-03T16:56:08Z"
  finalizers:
  - kubernetes.io/pv-protection
  name: fileserver
  resourceVersion: "56222"
  selfLink: /api/v1/persistentvolumes/fileserver
  uid: d50051e9-fe5a-11e9-842d-42010aa200e7
spec:
  accessModes:
  - ReadWriteMany
  capacity:
    storage: 1T
  claimRef:
    apiVersion: v1
    kind: PersistentVolumeClaim
    name: brigade-worker-01drs34caxbnmx2b8keafw9aqp
    namespace: default
    resourceVersion: "19845"
    uid: dfc5b205-fe5a-11e9-842d-42010aa200e7
  nfs:
    path: /brigadestorage
    server: 10.9.92.18
  persistentVolumeReclaimPolicy: Retain
  storageClassName: standard
  volumeMode: Filesystem
status:
  phase: Released
```

### Output of `brig version`:
v1.2.0

### Output of `kubectl version`:

Client Version: version.Info{Major:"1", Minor:"14", GitVersion:"v1.14.6", GitCommit:"96fac5cd13a5dc064f7d9f4f23030a6aeface6cc", GitTreeState:"clean", BuildDate:"2019-08-19T11:13:49Z", GoVersion:"go1.12.9", Compiler:"gc", Platform:"darwin/amd64"}
Server Version: version.Info{Major:"1", Minor:"13+", GitVersion:"v1.13.11-gke.9", GitCommit:"6c1e92d07f5717440f751666d4aad6943015d3cb", GitTreeState:"clean", BuildDate:"2019-10-11T23:14:17Z", GoVersion:"go1.11.13b4", Compiler:"gc", Platform:"linux/amd64"}

Cloud Provider/Platform (AKS, GKE, Minikube etc.):
GKE with filestorage for ReadWriteMany .

**Comment by krancour** [comment]
While perhaps not your desired behavior (nor mine, tbh), this is normal. For better or worse, job pods stick around after a build completes because logs from each job aren't stored anywhere else, so the continued existence of those pods is the only way that the logs from each job remain accessible. PVCs used by each of those pods don't complete termination until those pods are deleted.

This could change in the future if we do something about https://github.com/brigadecore/brigade/issues/960-- likely in a hypothetical 2.0 timeframe.

**Comment by technosophos** [comment]
I will look into this. The PVC should be destroyed as soon as the build itself is done. So possibly the teardown handler is not being run.

**Comment by technosophos** [comment]
Is the PVC deleted, but the PV remains? The output above shows the PV.

1. It is true that if a PVC is still around when a pod is deleted, the PVC will be deleted
2. However, we manually delete the PVC when the job is complete, and that should work (I assume, for example, that the individual build pods do not have their PVCs hanging around)

Here's the section of code that produces the last log message in the original issue (in k8s.ts):

```javascript
  public destroy(): Promise<boolean> {
    if (!this.proj && !this.name) {
      this.logger.log("Build storage not exists");
      return Promise.resolve(false);
    }

    this.logger.log(`Destroying PVC named ${this.name}`);
    let opts = new kubernetes.V1DeleteOptions();
    return Promise.resolve<boolean>(
      defaultClient
        .deleteNamespacedPersistentVolumeClaim(
          this.name,
          this.proj.kubernetes.namespace,
          "true",
          opts
        )
        .then(() => {
          return true;
        })
    );
  }
```

If the above throws an exception, then it is caught and logged one layer above. It would be useful to find a way to see if the PVC was marked for deletion (maybe with a `kubectl describe` or something). It _may_ be the case that the finalizer policy is preventing the PV from being deleted, though I'm not sure how one could do that.

But it may also be the case that somehow we're getting an error there that we are not trapping correctly. I'll keep digging.

**Comment by technosophos** [comment]
Also, I just noticed that the `phase` for your PV is `Released`, which should mean that it is unbound from all things that may have bound it. Can you tell me what you are backing this with? I'd like to go look at how they do cleanup of PVs.

**Comment by krancour** [comment]
> While perhaps not your desired behavior (nor mine, tbh), this is normal.

For the record, @technosophos is telling me now that I am wrong and this isn't normal.

What volume provisioner are you using? The issue _could_ lie therein. I'm using NFS and seeing the same behavior.

**Comment by bnadeem** [comment]
@krancour , The PV is connected to GCP Filestore via NFS.  

@technosophos , I'll recreate the scenario and get with more data.

**Comment by bnadeem** [comment]
After the build is finished, following are the results

### brig run log
```
Event created. Waiting for worker pod named "brigade-worker-01ds1a03cd22q4d54wfcc3qf9c".
Build: 01ds1a03cd22q4d54wfcc3qf9c, Worker: brigade-worker-01ds1a03cd22q4d54wfcc3qf9c
prestart: no dependencies file found
[brigade] brigade-worker version: 1.2.1
[brigade:k8s] Creating PVC named brigade-worker-01ds1a03cd22q4d54wfcc3qf9c
[brigade:k8s] Creating secret one-01ds1a03cd22q4d54wfcc3qf9c
[brigade:k8s] Creating pod one-01ds1a03cd22q4d54wfcc3qf9c
[brigade:k8s] Timeout set at 900000 milliseconds
[brigade:k8s] Pod not yet scheduled
[brigade:k8s] default/one-01ds1a03cd22q4d54wfcc3qf9c phase Pending
[brigade:k8s] default/one-01ds1a03cd22q4d54wfcc3qf9c phase Pending
done         
[brigade:k8s] default/one-01ds1a03cd22q4d54wfcc3qf9c phase Succeeded
[brigade:k8s] Creating secret two-01ds1a03cd22q4d54wfcc3qf9c
done         
[brigade:k8s] Creating pod two-01ds1a03cd22q4d54wfcc3qf9c
[brigade:k8s] Timeout set at 900000 milliseconds
[brigade:k8s] Pod not yet scheduled
[brigade:k8s] default/two-01ds1a03cd22q4d54wfcc3qf9c phase Pending
[brigade:k8s] default/two-01ds1a03cd22q4d54wfcc3qf9c phase Succeeded
[brigade:k8s] Creating secret three-01ds1a03cd22q4d54wfcc3qf9c
[brigade:k8s] Creating pod three-01ds1a03cd22q4d54wfcc3qf9c
[brigade:k8s] Timeout set at 900000 milliseconds
[brigade:k8s] Pod not yet scheduled
[brigade:k8s] default/three-01ds1a03cd22q4d54wfcc3qf9c phase Pending
[brigade:k8s] default/three-01ds1a03cd22q4d54wfcc3qf9c phase Succeeded
[brigade:app] after: default event handler fired
[brigade:app] beforeExit(2): destroying storage
[brigade:k8s] Destroying PVC named brigade-worker-01ds1a03cd22q4d54wfcc3qf9c
```

### PV

```
metadata:
  name: fileserver
  selfLink: /api/v1/persistentvolumes/fileserver
  uid: d2160a73-00dc-11ea-a1a0-42010aa2004c
  resourceVersion: '966653'
  creationTimestamp: '2019-11-06T21:31:40Z'
  annotations:
    kubectl.kubernetes.io/last-applied-configuration: >
      {"apiVersion":"v1","kind":"PersistentVolume","metadata":{"annotations":{},"name":"fileserver"},"spec":{"accessModes":["ReadWriteMany"],"capacity":{"storage":"1T"},"nfs":{"path":"/brigadestorage","server":"10.9.92.18"},"storageClassName":"standard"}}
    pv.kubernetes.io/bound-by-controller: 'yes'
  finalizers:
    - kubernetes.io/pv-protection
spec:
  capacity:
    storage: 1T
  nfs:
    server: 10.9.92.18
    path: /brigadestorage
  accessModes:
    - ReadWriteMany
  claimRef:
    kind: PersistentVolumeClaim
    namespace: default
    name: brigade-worker-01ds1a03cd22q4d54wfcc3qf9c
    uid: a72fd83d-00dc-11ea-a1a0-42010aa2004c
    apiVersion: v1
    resourceVersion: '966381'
  persistentVolumeReclaimPolicy: Retain
  storageClassName: standard
  volumeMode: Filesystem
status:
  phase: Bound
kind: PersistentVolume
```

### PVC

```
metadata:
  name: brigade-worker-01ds1a03cd22q4d54wfcc3qf9c
  namespace: default
  selfLink: >-
    /api/v1/namespaces/default/persistentvolumeclaims/brigade-worker-01ds1a03cd22q4d54wfcc3qf9c
  uid: a72fd83d-00dc-11ea-a1a0-42010aa2004c
  resourceVersion: '966737'
  creationTimestamp: '2019-11-06T21:30:28Z'
  deletionTimestamp: '2019-11-06T21:32:02Z'
  deletionGracePeriodSeconds: 0
  labels:
    build: 01ds1a03cd22q4d54wfcc3qf9c
    component: buildStorage
    heritage: brigade
    project: brigade-5e1ecee06a7fc06f305ae5c12acfe7a7f67b8ece7af76932ed3afa
    worker: brigade-worker-01ds1a03cd22q4d54wfcc3qf9c
  annotations:
    pv.kubernetes.io/bind-completed: 'yes'
    pv.kubernetes.io/bound-by-controller: 'yes'
    volume.beta.kubernetes.io/storage-provisioner: kubernetes.io/gce-pd
  finalizers:
    - kubernetes.io/pvc-protection
spec:
  accessModes:
    - ReadWriteMany
  resources:
    requests:
      storage: 50Mi
  volumeName: fileserver
  storageClassName: standard
  volumeMode: Filesystem
status:
  phase: Bound
  accessModes:
    - ReadWriteMany
  capacity:
    storage: 1T
kind: PersistentVolumeClaim
```

### kubectl describe pod `brigade-worker-01ds1a03cd22q4d54wfcc3qf9c`:

```
Name:               brigade-worker-01ds1a03cd22q4d54wfcc3qf9c
Namespace:          default
Priority:           0
PriorityClassName:  <none>
Node:               gke-brigade-default-pool-e3d03a50-qb7q/10.0.0.35
Start Time:         Wed, 06 Nov 2019 16:30:26 -0500
Labels:             build=01ds1a03cd22q4d54wfcc3qf9c
                    component=build
                    heritage=brigade
                    project=brigade-5e1ecee06a7fc06f305ae5c12acfe7a7f67b8ece7af76932ed3afa
Annotations:        kubernetes.io/limit-ranger: LimitRanger plugin set: cpu request for container brigade-runner
Status:             Succeeded
IP:                 10.60.0.149
Containers:
  brigade-runner:
    Container ID:  docker://882e901879d908b36f827735cf3f65fbdf8e65a22276b800b7b7b118c4eb5f49
    Image:         brigadecore/brigade-worker:v1.2.1
    Image ID:      docker-pullable://brigadecore/brigade-worker@sha256:9e5e2b263351249dfc57f2d61930e1c2835b1a76b8171e61ea5c0c8d0b2bc9bb
    Port:          <none>
    Host Port:     <none>
    Command:
      yarn
      -s
      start
    State:          Terminated
      Reason:       Completed
      Exit Code:    0
      Started:      Wed, 06 Nov 2019 16:30:27 -0500
      Finished:     Wed, 06 Nov 2019 16:32:02 -0500
    Ready:          False
    Restart Count:  0
    Requests:
      cpu:  100m
    Environment:
      CI:                                   true
      BRIGADE_BUILD_ID:                     01ds1a03cd22q4d54wfcc3qf9c
      BRIGADE_BUILD_NAME:                   brigade-worker-01ds1a03cd22q4d54wfcc3qf9c
      BRIGADE_COMMIT_ID:                    
      BRIGADE_COMMIT_REF:                   master
      BRIGADE_EVENT_PROVIDER:               brigade-cli
      BRIGADE_EVENT_TYPE:                   exec
      BRIGADE_PROJECT_ID:                   brigade-5e1ecee06a7fc06f305ae5c12acfe7a7f67b8ece7af76932ed3afa
      BRIGADE_LOG_LEVEL:                    log
      BRIGADE_REMOTE_URL:                   https://github.com/brigadecore/empty-testbed.git
      BRIGADE_WORKSPACE:                    /vcs
      BRIGADE_PROJECT_NAMESPACE:            default
      BRIGADE_SERVICE_ACCOUNT:              brigade-worker
      BRIGADE_SECRET_KEY_REF:               false
      BRIGADE_REPO_KEY:                     <set to the key 'sshKey' in secret 'brigade-5e1ecee06a7fc06f305ae5c12acfe7a7f67b8ece7af76932ed3afa'>        Optional: true
      BRIGADE_REPO_AUTH_TOKEN:              <set to the key 'github.token' in secret 'brigade-5e1ecee06a7fc06f305ae5c12acfe7a7f67b8ece7af76932ed3afa'>  Optional: true
      BRIGADE_DEFAULT_BUILD_STORAGE_CLASS:  
      BRIGADE_DEFAULT_CACHE_STORAGE_CLASS:  
      BRIGADE_SERVICE_ACCOUNT_REGEX:        brigade-worker
    Mounts:
      /etc/brigade from brigade-build (ro)
      /etc/brigade-project from brigade-project (ro)
      /var/run/secrets/kubernetes.io/serviceaccount from brigade-worker-token-lzk52 (ro)
Conditions:
  Type              Status
  Initialized       True 
  Ready             False 
  ContainersReady   False 
  PodScheduled      True 
Volumes:
  brigade-build:
    Type:        Secret (a volume populated by a Secret)
    SecretName:  brigade-worker-01ds1a03cd22q4d54wfcc3qf9c
    Optional:    false
  brigade-project:
    Type:        Secret (a volume populated by a Secret)
    SecretName:  brigade-5e1ecee06a7fc06f305ae5c12acfe7a7f67b8ece7af76932ed3afa
    Optional:    false
  brigade-worker-token-lzk52:
    Type:        Secret (a volume populated by a Secret)
    SecretName:  brigade-worker-token-lzk52
    Optional:    false
QoS Class:       Burstable
Node-Selectors:  beta.kubernetes.io/os=linux
Tolerations:     node.kubernetes.io/not-ready:NoExecute for 300s
                 node.kubernetes.io/unreachable:NoExecute for 300s
Events:
  Type    Reason     Age   From                                                       Message
  ----    ------     ----  ----                                                       -------
  Normal  Scheduled  16m   default-scheduler                                          Successfully assigned default/brigade-worker-01ds1a03cd22q4d54wfcc3qf9c to gke-brigade-default-pool-e3d03a50-qb7q
  Normal  Pulled     16m   kubelet, gke-brigade-default-pool-e3d03a50-qb7q  Container image "brigadecore/brigade-worker:v1.2.1" already present on machine
  Normal  Created    16m   kubelet, gke-brigade-default-pool-e3d03a50-qb7q  Created container
  Normal  Started    16m   kubelet, gke-brigade-default-pool-e3d03a50-qb7q  Started container


```

### kubectl describe pod one-01ds1a03cd22q4d54wfcc3qf9c

```
Name:               one-01ds1a03cd22q4d54wfcc3qf9c
Namespace:          default
Priority:           0
PriorityClassName:  <none>
Node:               gke-brigade-default-pool-e3d03a50-83b1/10.0.0.36
Start Time:         Wed, 06 Nov 2019 16:31:44 -0500
Labels:             build=01ds1a03cd22q4d54wfcc3qf9c
                    component=job
                    heritage=brigade
                    jobname=one
                    project=brigade-5e1ecee06a7fc06f305ae5c12acfe7a7f67b8ece7af76932ed3afa
                    worker=brigade-worker-01ds1a03cd22q4d54wfcc3qf9c
Annotations:        kubernetes.io/limit-ranger: LimitRanger plugin set: cpu request for container brigaderun
Status:             Succeeded
IP:                 10.60.1.26
Containers:
  brigaderun:
    Container ID:  docker://f63879ea1cee807020a3d1835057f5637ec863a5566f901439f483784315809e
    Image:         alpine:3.4
    Image ID:      docker-pullable://alpine@sha256:0325f4ff0aa8c89a27d1dbe10b29a71a8d4c1a42719a4170e0552a312e22fe88
    Port:          <none>
    Host Port:     <none>
    Command:
      /bin/sh
      /hook/main.sh
    State:          Terminated
      Reason:       Completed
      Exit Code:    0
      Started:      Wed, 06 Nov 2019 16:31:48 -0500
      Finished:     Wed, 06 Nov 2019 16:31:48 -0500
    Ready:          False
    Restart Count:  0
    Requests:
      cpu:        100m
    Environment:  <none>
    Mounts:
      /hook from one-01ds1a03cd22q4d54wfcc3qf9c (rw)
      /mnt/brigade/share from build-storage (rw)
      /var/run/secrets/kubernetes.io/serviceaccount from brigade-worker-token-lzk52 (ro)
Conditions:
  Type              Status
  Initialized       True 
  Ready             False 
  ContainersReady   False 
  PodScheduled      True 
Volumes:
  one-01ds1a03cd22q4d54wfcc3qf9c:
    Type:        Secret (a volume populated by a Secret)
    SecretName:  one-01ds1a03cd22q4d54wfcc3qf9c
    Optional:    false
  build-storage:
    Type:       PersistentVolumeClaim (a reference to a PersistentVolumeClaim in the same namespace)
    ClaimName:  brigade-worker-01ds1a03cd22q4d54wfcc3qf9c
    ReadOnly:   false
  brigade-worker-token-lzk52:
    Type:        Secret (a volume populated by a Secret)
    SecretName:  brigade-worker-token-lzk52
    Optional:    false
QoS Class:       Burstable
Node-Selectors:  <none>
Tolerations:     node.kubernetes.io/not-ready:NoExecute for 300s
                 node.kubernetes.io/unreachable:NoExecute for 300s
Events:
  Type     Reason            Age                From                                                       Message
  ----     ------            ----               ----                                                       -------
  Warning  FailedScheduling  14m (x6 over 15m)  default-scheduler                                          pod has unbound immediate PersistentVolumeClaims (repeated 3 times)
  Normal   Scheduled         14m                default-scheduler                                          Successfully assigned default/one-01ds1a03cd22q4d54wfcc3qf9c to gke-brigade-default-pool-e3d03a50-83b1
  Normal   Pulled            14m                kubelet, gke-brigade-default-pool-e3d03a50-83b1  Container image "alpine:3.4" already present on machine
  Normal   Created           14m                kubelet, gke-brigade-default-pool-e3d03a50-83b1  Created container
  Normal   Started           14m                kubelet, gke-brigade-default-pool-e3d03a50-83b1  Started container

```
### kubectl describe pod two-01ds1a03cd22q4d54wfcc3qf9c

```
Name:               two-01ds1a03cd22q4d54wfcc3qf9c
Namespace:          default
Priority:           0
PriorityClassName:  <none>
Node:               gke-brigade-default-pool-e3d03a50-83b1/10.0.0.36
Start Time:         Wed, 06 Nov 2019 16:31:50 -0500
Labels:             build=01ds1a03cd22q4d54wfcc3qf9c
                    component=job
                    heritage=brigade
                    jobname=two
                    project=brigade-5e1ecee06a7fc06f305ae5c12acfe7a7f67b8ece7af76932ed3afa
                    worker=brigade-worker-01ds1a03cd22q4d54wfcc3qf9c
Annotations:        kubernetes.io/limit-ranger: LimitRanger plugin set: cpu request for container brigaderun
Status:             Succeeded
IP:                 10.60.1.27
Containers:
  brigaderun:
    Container ID:  docker://2ec1ebfc38188412c1c2e77b6fc1e8f04d20c683c2198250e884200623b0ba14
    Image:         alpine:3.4
    Image ID:      docker-pullable://alpine@sha256:0325f4ff0aa8c89a27d1dbe10b29a71a8d4c1a42719a4170e0552a312e22fe88
    Port:          <none>
    Host Port:     <none>
    Command:
      /bin/sh
      /hook/main.sh
    State:          Terminated
      Reason:       Completed
      Exit Code:    0
      Started:      Wed, 06 Nov 2019 16:31:55 -0500
      Finished:     Wed, 06 Nov 2019 16:31:55 -0500
    Ready:          False
    Restart Count:  0
    Requests:
      cpu:        100m
    Environment:  <none>
    Mounts:
      /hook from two-01ds1a03cd22q4d54wfcc3qf9c (rw)
      /mnt/brigade/share from build-storage (rw)
      /var/run/secrets/kubernetes.io/serviceaccount from brigade-worker-token-lzk52 (ro)
Conditions:
  Type              Status
  Initialized       True 
  Ready             False 
  ContainersReady   False 
  PodScheduled      True 
Volumes:
  two-01ds1a03cd22q4d54wfcc3qf9c:
    Type:        Secret (a volume populated by a Secret)
    SecretName:  two-01ds1a03cd22q4d54wfcc3qf9c
    Optional:    false
  build-storage:
    Type:       PersistentVolumeClaim (a reference to a PersistentVolumeClaim in the same namespace)
    ClaimName:  brigade-worker-01ds1a03cd22q4d54wfcc3qf9c
    ReadOnly:   false
  brigade-worker-token-lzk52:
    Type:        Secret (a volume populated by a Secret)
    SecretName:  brigade-worker-token-lzk52
    Optional:    false
QoS Class:       Burstable
Node-Selectors:  <none>
Tolerations:     node.kubernetes.io/not-ready:NoExecute for 300s
                 node.kubernetes.io/unreachable:NoExecute for 300s
Events:
  Type    Reason     Age   From                                                       Message
  ----    ------     ----  ----                                                       -------
  Normal  Scheduled  15m   default-scheduler                                          Successfully assigned default/two-01ds1a03cd22q4d54wfcc3qf9c to gke-brigade-default-pool-e3d03a50-83b1
  Normal  Pulled     15m   kubelet, gke-brigade-default-pool-e3d03a50-83b1  Container image "alpine:3.4" already present on machine
  Normal  Created    15m   kubelet, gke-brigade-default-pool-e3d03a50-83b1  Created container
  Normal  Started    15m   kubelet, gke-brigade-default-pool-e3d03a50-83b1  Started container

```

### kubectl describe pod three-01ds1a03cd22q4d54wfcc3qf9c:
```
Name:               three-01ds1a03cd22q4d54wfcc3qf9c
Namespace:          default
Priority:           0
PriorityClassName:  <none>
Node:               gke-brigade-default-pool-e3d03a50-83b1/10.0.0.36
Start Time:         Wed, 06 Nov 2019 16:31:56 -0500
Labels:             build=01ds1a03cd22q4d54wfcc3qf9c
                    component=job
                    heritage=brigade
                    jobname=three
                    project=brigade-5e1ecee06a7fc06f305ae5c12acfe7a7f67b8ece7af76932ed3afa
                    worker=brigade-worker-01ds1a03cd22q4d54wfcc3qf9c
Annotations:        kubernetes.io/limit-ranger: LimitRanger plugin set: cpu request for container brigaderun
Status:             Succeeded
IP:                 10.60.1.28
Containers:
  brigaderun:
    Container ID:  docker://747dac29f583ff5ef1957725b1855dc79528364972d0b64c2f3b4f89950d5190
    Image:         alpine:3.4
    Image ID:      docker-pullable://alpine@sha256:0325f4ff0aa8c89a27d1dbe10b29a71a8d4c1a42719a4170e0552a312e22fe88
    Port:          <none>
    Host Port:     <none>
    Command:
      /bin/sh
      /hook/main.sh
    State:          Terminated
      Reason:       Completed
      Exit Code:    0
      Started:      Wed, 06 Nov 2019 16:32:01 -0500
      Finished:     Wed, 06 Nov 2019 16:32:01 -0500
    Ready:          False
    Restart Count:  0
    Requests:
      cpu:        100m
    Environment:  <none>
    Mounts:
      /hook from three-01ds1a03cd22q4d54wfcc3qf9c (rw)
      /mnt/brigade/share from build-storage (rw)
      /var/run/secrets/kubernetes.io/serviceaccount from brigade-worker-token-lzk52 (ro)
Conditions:
  Type              Status
  Initialized       True 
  Ready             False 
  ContainersReady   False 
  PodScheduled      True 
Volumes:
  three-01ds1a03cd22q4d54wfcc3qf9c:
    Type:        Secret (a volume populated by a Secret)
    SecretName:  three-01ds1a03cd22q4d54wfcc3qf9c
    Optional:    false
  build-storage:
    Type:       PersistentVolumeClaim (a reference to a PersistentVolumeClaim in the same namespace)
    ClaimName:  brigade-worker-01ds1a03cd22q4d54wfcc3qf9c
    ReadOnly:   false
  brigade-worker-token-lzk52:
    Type:        Secret (a volume populated by a Secret)
    SecretName:  brigade-worker-token-lzk52
    Optional:    false
QoS Class:       Burstable
Node-Selectors:  <none>
Tolerations:     node.kubernetes.io/not-ready:NoExecute for 300s
                 node.kubernetes.io/unreachable:NoExecute for 300s
Events:
  Type    Reason     Age   From                                                       Message
  ----    ------     ----  ----                                                       -------
  Normal  Scheduled  16m   default-scheduler                                          Successfully assigned default/three-01ds1a03cd22q4d54wfcc3qf9c to gke-brigade-default-pool-e3d03a50-83b1
  Normal  Pulled     16m   kubelet, gke-brigade-default-pool-e3d03a50-83b1  Container image "alpine:3.4" already present on machine
  Normal  Created    16m   kubelet, gke-brigade-default-pool-e3d03a50-83b1  Created container
  Normal  Started    16m   kubelet, gke-brigade-default-pool-e3d03a50-83b1  Started container
```

### After deleting the job pods ( one-01ds1a03cd22q4d54wfcc3qf9c, three-01ds1a03cd22q4d54wfcc3qf9c, two-01ds1a03cd22q4d54wfcc3qf9c )

PVC is deleted and PV is in Released status.

### PV after pods created for jobs are deleted: 

```
metadata:
  name: fileserver
  selfLink: /api/v1/persistentvolumes/fileserver
  uid: d2160a73-00dc-11ea-a1a0-42010aa2004c
  resourceVersion: '970396'
  creationTimestamp: '2019-11-06T21:31:40Z'
  annotations:
    kubectl.kubernetes.io/last-applied-configuration: >
      {"apiVersion":"v1","kind":"PersistentVolume","metadata":{"annotations":{},"name":"fileserver"},"spec":{"accessModes":["ReadWriteMany"],"capacity":{"storage":"1T"},"nfs":{"path":"/brigadestorage","server":"10.9.92.18"},"storageClassName":"standard"}}
    pv.kubernetes.io/bound-by-controller: 'yes'
  finalizers:
    - kubernetes.io/pv-protection
spec:
  capacity:
    storage: 1T
  nfs:
    server: 10.9.92.18
    path: /brigadestorage
  accessModes:
    - ReadWriteMany
  claimRef:
    kind: PersistentVolumeClaim
    namespace: default
    name: brigade-worker-01ds1a03cd22q4d54wfcc3qf9c
    uid: a72fd83d-00dc-11ea-a1a0-42010aa2004c
    apiVersion: v1
    resourceVersion: '966381'
  persistentVolumeReclaimPolicy: Retain
  storageClassName: standard
  volumeMode: Filesystem
status:
  phase: Released
kind: PersistentVolume
```

### kubectl describe pv fileserver

```
Name:            fileserver
Labels:          <none>
Annotations:     kubectl.kubernetes.io/last-applied-configuration:
                   {"apiVersion":"v1","kind":"PersistentVolume","metadata":{"annotations":{},"name":"fileserver"},"spec":{"accessModes":["ReadWriteMany"],"ca...
                 pv.kubernetes.io/bound-by-controller: yes
Finalizers:      [kubernetes.io/pv-protection]
StorageClass:    standard
Status:          Released
Claim:           default/brigade-worker-01ds1a03cd22q4d54wfcc3qf9c
Reclaim Policy:  Retain
Access Modes:    RWX
VolumeMode:      Filesystem
Capacity:        1T
Node Affinity:   <none>
Message:         
Source:
    Type:      NFS (an NFS mount that lasts the lifetime of a pod)
    Server:    10.9.92.18
    Path:      /brigadestorage
    ReadOnly:  false
Events:        <none>
```

**Comment by technosophos** [comment]
Okay. I am going to do some testing. I am suspecting that there is a bug in the clean-up logic for the build-wide PVC because the per-Job PVCs seem to be getting cleaned up (right?).

For example, I see:

```
Every 2.0s: kubectl get pvc                                                                                   ts: Wed Nov  6 20:07:28 2019

NAME                                                             STATUS   VOLUME                                     CAPACITY   ACCESS MOD
ES   STORAGECLASS   AGE
brigade-aea88cfd4231bc98fff08a51a6dad9256eb0f684fa3e9f0556a5a3   Bound    pvc-74ef2762-fce1-4efc-8d55-efb332d684d2   50Mi       RWX
     standard       16s
github-com-technosophos-buck-cnab-porter-run                     Bound    pvc-db6d1939-e3a7-4f9b-a9c2-aace0697fde2   20Mi       RWX
     standard       26h

```

but after the build the `brigade-aea88cfd4231bc98fff08a51a6dad9256eb0f684fa3e9f0556a5a3` PVC is torn down.

I was using `watch kubectl get pvc` while the job was running so I could watch the PVCs get created and destroyed.

I'm going to do some more tests tomorrow. Thank you so much for the thorough research @bnadeem . That helps a lot.

**Comment by technosophos** [comment]
I am able to reproduce this on AKS with the `stable` helm chart for `nfs-server-provisioner`. What I see in the NFS provisioner case is that the PVC will stay in `terminating` indefinitely long -- probably until the pod itself is deleted.

The PVC is cleaned up immediately if I remove `storage.enabled`, but if it is mounted (which is what `storage.enabled` does), then the PVC stays in `terminating`. So @krancour is right: the PVC hangs around for as long as the pod's record lives. To me, that seems like a bug in the PVC system, though, since this would be true for _all_ pods, and is not uniquely a Brigade problem. Brigade does nothing special to hold a job there after it is complete. That's just the way Kubernetes works.

So at this point I feel like I need someone with more PVC knowledge to weigh in and tell us if there is some way to override this behavior, perhaps through a finalizer policy.

**Comment by technosophos** [comment]
I just noticed the following callout in the [PV documentation](https://kubernetes.io/docs/concepts/storage/persistent-volumes/#storage-object-in-use-protection):

> PVC is in active use by a Pod when a Pod object exists that is using the PVC.

That seems to imply that the finalizer will prevent deletion of the PVC (and hence the PV) for as long as the Pod record exists in Kubernetes. So perhaps being more aggressive about deleting completed pods is the solution? In Brigade, you can turn up the frequency of the vacuum.

**Comment by technosophos** [comment]
This morning some colleagues pointed me to similar issues all around the Kubernetes ecosystem.

tl;dr: The proposed solution is to delete the finalizer off of the object once the object goes into `status: Terminating`. Is this something we want to consider? (It's a little heavy-handed)

References:

https://github.com/kubernetes/kubernetes/issues/60807#issuecomment-401039268
https://github.com/kubernetes/kubernetes/issues/60807#issuecomment-518359503
https://github.com/kubernetes/kubernetes/issues/60807#issuecomment-546844175
https://github.com/thyarles/knsk

**Comment by kenotsolutions** [comment]
I see the same problem when I use openebs, It opens some "pods" in openebs namespace and brigade vacuum does not delete them.

**Comment by krancour** [comment]
Closing due to staleness. Please do re-open if appropriate.
